## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
print(df.columns)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [4]:
# Use the describe function to gather some basic statistics
print(df.describe())


          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.



In [7]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().head(3)
print(top_categories)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [11]:
# For the category with the most entries,
# which subcategory had the most entries?
# Find the category with the most entries
most_common_category = df['category'].value_counts().idxmax()

# Filter the DataFrame to include only rows with the most common category
filtered_df = df[df['category'] == most_common_category]

# Find the subcategory with the most entries within the filtered DataFrame
most_common_subcategory = filtered_df['subcategory'].value_counts().idxmax()

# Count the number of items in the most common subcategory
num_items_subcategory = filtered_df['subcategory'].value_counts().max()

#print(f"The subcategory with the most entries in the category '{most_common_category}' is '{most_common_subcategory}' with '{num_items_subcategory}' items.")

print(f'{most_common_category} {most_common_subcategory} {num_items_subcategory}')



The subcategory with the most entries in the category 'consumables' is 'bathroom supplies' with '6424' items.
consumables bathroom supplies 6424


In [17]:
# Which five clients had the most entries in the data?
top_five_clients = df['client_id'].value_counts().head(5)
top_five_clients


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [19]:
# Store the client ids of those top 5 clients in a list.
top_five_list = top_five_clients.index.tolist()
top_five_list

[33615, 66037, 46820, 38378, 24741]

In [24]:
# How many total units (the qty column) did the
# client with the most entries order order?
most_common_client = df['client_id'].value_counts().idxmax()

# Filtering the DataFrame to only include rows for the most common client
filtered_df = df[df['client_id'] == most_common_client]

# Calculating the total units
total_units_ordered = filtered_df['qty'].sum()

total_units_ordered

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [29]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']

print(df[['unit_price' , 'qty' , 'line_subtotal']].head())


   unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95
2       13.52   39         527.28
3       36.42   29        1056.18
4      195.10   20        3902.00


In [40]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound 
# for orders over 50 pounds and $10 per 
# pound for items 50 pounds or under.

def calc_ship_price(row):
    if row['total_weight'] <= 50:
        return row['total_weight'] * 10 # if shipping is less than 50 pounts, then its $10
    else:
        return row['total_weight'] * 7 # shipping more than 50 pounds, then its $7

# Instructions stated to create a column for shipping price, 
# however the listed results in the notebook had a colum for total_weight.
# Therefore, created an additional column to mimic previous results.
# Called it 'total_weight' to match name.
        
def total_weight(row):
    return row['unit_weight'] * row['qty']

df['total_weight'] = df.apply(total_weight, axis = 1)

# Created column for shipping prices.

df['shipping_price'] = df.apply(calc_ship_price, axis = 1)

print(df[['unit_price' , 'unit_weight' , 'qty' , 'total_weight' , 'shipping_price']].head(3))


   unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64


In [41]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

# Defining a function to calculate total price
def calc_total_price (row):
    subtotal = row['line_subtotal']
    shipping_price = row['shipping_price']
    # sales_tax_rate = 0.0925 

    line_price = subtotal + shipping_price
    total_price_with_tax = line_price + (line_price * 0.0925)

    return round(total_price_with_tax, 2) # rounding answer to two decimal places

df['line_price'] = df.apply(calc_total_price, axis = 1)

#Priting the DataFrame columns to match given answer
print(df[['line_subtotal' , 'shipping_price' , 'line_price']].head(3))


   line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12


In [49]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

df['line_subtotal'] = df['unit_cost'] * df['qty']

df['line_cost'] = df['line_subtotal'] + df['shipping_price']

# Created a list to specify what columns I want to show in the answer

columns_to_display = [
    'first'         , 
    'last'          , 
    'job'           , 
    'phone'         , 
    'email'         , 
    'client_id'     , 
    'order_id'      ,
    'order_date'    , 
    'order_week'    , 
    'order_year'    , 
    'unit_price'    , 
    'unit_cost'     ,
    'unit_weight'   , 
    'qty'           , 
    'line_number'   , 
    'line_subtotal' , 
    'total_weight'  ,
    'shipping_price', 
    'line_price'    , 
    'line_cost'
    ]

# Printing columns 
print(df[columns_to_display].head(3).to_string(index=False, float_format='%.2f'))

  first    last             job              phone                             email  client_id  order_id order_date  order_week  order_year  unit_price  unit_cost  unit_weight  qty  line_number  line_subtotal  total_weight  shipping_price  line_price  line_cost
 Donald Harding    Immunologist 793-904-7725x39308  harding.donald.7185@sullivan.com      58515   8953482 2023-04-28          17        2023     1096.80     762.71         7.50  105            1       80084.55        787.50         5512.50   131839.08   85597.05
Tiffany   Myers Music therapist   201.442.4543x942 myers.t.6537@ferguson-johnson.net      37609   8069089 2023-05-19          20        2023       24.95      15.09         1.49   21            0         316.89         31.29          312.90      914.26     629.79
Shannon  Watson    Immunologist  687.737.9424x8503             swatson8146@payne.net      57113   1902144 2023-01-29           4        2023       13.52       7.86         1.68   39            6         306.54  

In [14]:
# Create a column for the profit of
# each line using line cost and line price



first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0      115164.00        787.50         5512.50   131839.08   85597.05   
1         523.95         31.29          312.90      914.26     629.79   
2         527.28         65.52          458.64     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above


Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.



33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02


In [17]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.



client_id     qty  shipping_price   line_price    line_cost  line_profit
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61

In [18]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.



In [19]:
# Sort the updated data by "Total Profit" form highest to lowest


Client ID   Units Shipping Total Revenue Total Cost Total Profit
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
1      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
2      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
0      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M